### PR TITLE
Updated DAC Validation Procedure to Match the Spec.

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
The attestation certificate chain validation doesn't match the spec requirements.
ticket #21007 

#### Change overview
  - Removed use of IsCertificateValidAtCurrentTime() as spec doesn't require
    DAC certificate validity check at the current time.
  - Updated mbedTLS implementation of the ValidateCertificateChain() to ignore
    validity checks of the certificates in the validation chain.
  - Updated OpenSSL implementation of the ValidateCertificateChain() to use
    notBefore time of the DAC certificate as a current time. That way the chain
    validation function checks exactly that the PAA and PAI were valid at the
    time of DAC generation.
  - NOTE: Ideally it would be nice if mbedTLS and OpenSSL implementations of
    ValidateCertificateChain() function behaive similar. Unfortunately, I didn't
    find a way to adjust the mbedTLS current time during the chain validation.
  - Created new test vectors for PAA, PAI, and DAC (correct and with errors)
    to validate all various corner cases of the implementation.

#### Testing
Generated new test vectors and added new tests

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 1]
```diff
@@ -1252,6 +1252,14 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointIsValid(void * R)
     return CHIP_NO_ERROR;
 }
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+static int mbedTLSCallbackIgnoreCertValidityCheck(void * data, mbedtls_x509_crt * crt, int depth, uint32_t * flags)
+{
+    *flags &= ~(static_cast<uint32_t>(MBEDTLS_X509_BADCERT_EXPIRED | MBEDTLS_X509_BADCERT_FUTURE));
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:1258]: Could this not parse the mbedTLS cert and use the depth to validate the rule, ratther than ignore the outcome?

**Review by tcarmelveilleux — APPROVED** [review]

File: src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp [hunk 2]
```diff
@@ -225,11 +225,6 @@ void DefaultDACVerifier::VerifyAttestationInformation(const DeviceAttestationVer
         VerifyOrExit(!paaVidPid.mProductId.HasValue(), attestationError = AttestationVerificationResult::kPaaFormatInvalid);
     }
 
-#if !defined(CURRENT_TIME_NOT_IMPLEMENTED)
-    VerifyOrExit(IsCertificateValidAtCurrentTime(info.dacDerBuffer) == CHIP_NO_ERROR,
-                 attestationError = AttestationVerificationResult::kDacExpired);
-#endif
-
```
Comment (msandstedt) [inline src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:232]: I had read the spec as still requiring this:

```
Given a list of DER-encoded certificates in RFC 5280 format, starting at the end-entity (leaf)
certificate, and following the chain of trust towards the root, Crypto_VerifyChainDER() performs all
the necessary validity checks on certificates.
The Validity period validation for the root and optional intermediate certificates is performed
against the notBefore timestamp of the end-entity (leaf certificate) used as value for the current
time.
```

Only PAA and PAI seem to be explicitly specified has having different rules for Validity period validation.

**Review by msandstedt — COMMENTED** [review]
Most of this looks functionally good to me, but it would be nice if we could close on this:

https://github.com/project-chip/connectedhomeip/pull/21531/files#r936068987

The operative question is whether a DAC requires NotAfter to be set to X509NoWellDefinedExpirationDate in order to achieve non-expiration for the chain.  The spec is very explicit that PAA and PAI don't require this, but I can't find any language that would remove the requirement for a DAC.

File: src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp (hunk 2 above)
Comment (emargolis) [reply src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:232]: Analyzing this section of the spec I agree with you that the DAC must be valid at the current time. So I am returning that check. And then I also agree with your second statement below that the DAC NotAfter should be set to X509NoWellDefinedExpirationDate in order to achieve non-expiration for the chain.
The spec should be updated to require one of the following:
  - require non-expiration for the DAC
  - or don't check validity of the DAC

File: src/crypto/CHIPCryptoPALmbedTLS.cpp (hunk 1 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALmbedTLS.cpp:1258]: I'll do it as a follow up: will see if that code should be restructured.

Comment (turon) [reply src/crypto/CHIPCryptoPALmbedTLS.cpp:1258]: Could you file an issue and add `// TODO(#issue): ` here to track this future work?

**Review by msandstedt — APPROVED** [review]

File: src/crypto/CHIPCryptoPALmbedTLS.cpp (hunk 1 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALmbedTLS.cpp:1258]: Opened ticker #21621

**Pull request merged**